### PR TITLE
A more principled API for exchanging container through the interface

### PR DESCRIPTION
**Pull request by pleroy**
Keeping `LineSegment` for the time being, will be removed in a later PR.

### CODE REVIEW

File: ksp_plugin/interface.hpp [hunk 1]
```diff
@@ -25,13 +26,45 @@ using ksp_plugin::World;
...
+  virtual ~Iterator() = default;
+
+  virtual bool AtEnd() const = 0;
+  virtual void Increment() = 0;
+  virtual int Size() const = 0;
 };
 
+// A concrete, typed subclass of |Iterator|.  It holds a |Container| of elements
+// of type |Internal|.
+template<typename Internal, template<typename...> class Container>
```
Comment (eggrobin) [inline ksp_plugin/interface.hpp]: Remove the template parameter `Internal`, and use `Container::value_type` (from the concept [`Container`](http://en.cppreference.com/w/cpp/concept/Container)) instead.
This should allow for type inference on the constructor, turning
```cpp
*periapsides = new TypedIterator<LineSegment<World>, std::vector>(std::move(rendered_periapsides));
```
into

```cpp
*periapsides = new TypedIterator(std::move(rendered_periapsides));
```

File: ksp_plugin_test/interface_test.cpp [hunk 2]
```diff
@@ -555,30 +555,30 @@ TEST_F(InterfaceTest, RenderedPrediction) {
...
   // Delete it.
-  EXPECT_THAT(line_and_iterator, Not(IsNull()));
-  principia__DeleteLineAndIterator(&line_and_iterator);
-  EXPECT_THAT(line_and_iterator, IsNull());
+  EXPECT_THAT(iterator, Not(IsNull()));
+  principia__IteratorDelete(&iterator);
+  EXPECT_THAT(iterator, IsNull());
 }
 
 TEST_F(InterfaceTest, LineAndIterator) {
```
Comment (eggrobin) [inline ksp_plugin_test/interface_test.cpp]: The name of this test should change.

File: ksp_plugin/interface_iterator.cpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+void principia__IteratorIncrement(Iterator* const iterator) {
+  journal::Method<journal::IteratorIncrement> m({iterator});
+  CHECK_NOTNULL(iterator)->Increment();
+  return m.Return();
+}
+
+XYZSegment principia__IteratorGetXYZSegment(Iterator const* const iterator) {
+  journal::Method<journal::IteratorGetXYZSegment> m({iterator});
+  CHECK_NOTNULL(iterator);
+  auto const* typed_iterator =
```
Comment (eggrobin) [inline ksp_plugin/interface_iterator.cpp]: `not_null<TypedIterator<std::vector<LineSegment<World>> const*> const`, to detect `dynamic_cast` failure.

File: ksp_plugin/interface_body.hpp [hunk 4]
```diff
@@ -9,10 +9,58 @@
 namespace principia {
 namespace interface {
 
+namespace {
```
Comment (eggrobin) [inline ksp_plugin/interface_body.hpp]: `internal`, this is a header.

File: ksp_plugin/interface.hpp (hunk 1 above)
Comment (pleroy) [inline ksp_plugin/interface.hpp]: Done.

File: ksp_plugin_test/interface_test.cpp (hunk 2 above)
Comment (pleroy) [inline ksp_plugin_test/interface_test.cpp]: Done.

File: ksp_plugin/interface_iterator.cpp (hunk 3 above)
Comment (pleroy) [inline ksp_plugin/interface_iterator.cpp]: Done.

File: ksp_plugin/interface_body.hpp (hunk 4 above)
Comment (pleroy) [inline ksp_plugin/interface_body.hpp]: Not today.

File: ksp_plugin/interface.hpp [hunk 5]
```diff
@@ -38,23 +38,25 @@ class Iterator {
 
 // A concrete, typed subclass of |Iterator|.  It holds a |Container| of elements
 // of type |Internal|.
```
Comment (eggrobin) [inline ksp_plugin/interface.hpp]: No such type.

**Pull request merged**